File: 2022/day05/go/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
	"time"
)

type crate []string
type Crate [9]crate

type instruction struct {
	n   int
	src int
	dst int
}

type Instructions []instruction

func loadData() (Crate, Instructions) {

	bs, _ := ioutil.ReadFile("../input.txt")
	input := strings.Split(string(bs), "\n")

	var crates Crate
	for i := 7; i >= 0; i-- {
		line := input[i]
		for j := 1; j < len(line); j += 4 {
			char := line[j : j+1]

			if char != " " {
				crateIdx := j / 4
				crates[crateIdx] = append(crates[crateIdx], char)
			}
		}
	}

	var instructions Instructions
	for i := 10; i < len(input); i++ {
		line := input[i]

		r := regexp.MustCompile(`\d+`)
		matches := r.FindAllString(line, -1)

		n, _ := strconv.Atoi(matches[0])
		src, _ := strconv.Atoi(matches[1])
		dst, _ := strconv.Atoi(matches[2])

		instr := instruction{n: n, src: src, dst: dst}

		instructions = append(instructions, instr)
	}

	return crates, instructions

}

func p1(crates Crate, instructions Instructions) []string {
	for _, instr := range instructions {
		for i := 0; i < instr.n; i++ {
			pop := crates[instr.src-1][len(crates[instr.src-1])-1]
			crates[instr.src-1] = crates[instr.src-1][:len(crates[instr.src-1])-1]
			crates[instr.dst-1] = append(crates[instr.dst-1], pop)
		}
	}

	output := []string{}
	for i := 0; i < len(crates); i++ {
		x := crates[i][len(crates[i])-1]
		output = append(output, x)
	}

	return output
}

func p2(crates Crate, instructions Instructions) []string {
	for _, instr := range instructions {
		items := []string{}
		for i := 0; i < instr.n; i++ {
			pop := crates[instr.src-1][len(crates[instr.src-1])-1]
			crates[instr.src-1] = crates[instr.src-1][:len(crates[instr.src-1])-1]
			items = append(items, pop)
		}

		for i := 0; i < instr.n; i++ {
			pop := items[len(items)-1]
			items = items[:len(items)-1]
			crates[instr.dst-1] = append(crates[instr.dst-1], pop)
		}
	}

	output := []string{}
	for i := 0; i < len(crates); i++ {
		x := crates[i][len(crates[i])-1]
		output = append(output, x)
	}

	return output
}

func main() {
	start := time.Now()

	crates, instructions := loadData()
	fmt.Println("The solution to part 1 is: ", p1(crates, instructions))

	crates, instructions = loadData()
	fmt.Println("The solution to part 2 is: ", p2(crates, instructions))

	end := time.Since(start)
	fmt.Println("Time: ", end)

}

// go run ../go/main.go

File: 2022/day05/nim/main.nim
import std/strutils
import std/[times]
import re


type
    Crate = seq[char]
    Crates = array[9, Crate]

    Instruction = object
        n: int
        src: int
        dst: int

    Instructions = seq[Instruction]


proc loadData(): (Crates, Instructions) = 
    let data = readFile("../input.txt").splitLines()

    var crates: Crates

    for i in countdown(7, 0):
        let line = data[i]
        for col in countup(1, 35, 4):
            let crateIdx = col div 4
            let val = line[col]
            if val != ' ':
                crates[crateIdx].add(val)

    var instructions: Instructions

    for i in countup(10, data.high):
        let ints = findAll(data[i], re"\d+", start=1)

        let instr = Instruction(
            n : parseInt(ints[0]),
            src : parseInt(ints[1]),
            dst : parseInt(ints[2])
        )

        instructions.add(instr)

    return (crates, instructions)



proc p1(crates: var Crates, instructions: Instructions): seq[char] =
    for instr in instructions:
        for i in 0..instr.n - 1:
            crates[instr.dst - 1].add(pop(crates[instr.src - 1]))
    
    var output: seq[char]
    for i in crates:
        output.add(i[^1])

    return output


proc p2(crates: var Crates, instructions: Instructions): seq[char] =
    for instr in instructions:
        
        var items: seq[char]
        for i in 0..instr.n - 1:
            items.add(pop(crates[instr.src - 1]))

        for i in 0..instr.n - 1:
            crates[instr.dst - 1].add(pop(items))

    var output: seq[char]
    for i in crates:
        output.add(i[^1])

    return output


let t0 = epochTime()

var (crates, instructions) = loadData()
echo "The solution to part 1 is: ", p1(crates, instructions)

(crates, instructions) = loadData()
echo "The solution to part 2 is: ", p2(crates, instructions)

let elapsed = (epochTime() - t0) * 1000
let elapsedStr = elapsed.formatFloat(format = ffDecimal, precision = 3)
echo "CPU Time: ", elapsedStr, "ms"

# nim c -r -d:release -d:danger main.nim